accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,577 @@
+- Efficient implementation of tying both `IDynamicMetaObjectProvider` and the writable DOM classes into a single instance. Newtonsoft also has a single object model with [JToken](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_Linq_JToken.htm) which implements `IDynamicMetaObjectProvider`.
+
+## Adding a System.Text.Json.Dynamic.dll (STJD) assembly
+Having `System.Text.Json.dll` directly reference `System.Linq.Expressions.dll` is not desired:
+- Implementations of Blazor and stand-alone apps that do not use `dynamic` features do not want to carry the large `System.Linq.Expressions.dll`.
+  - Referencing `System.Linq.Expressions.dll` will require ~5.5MB of disk size. `System.Linq.Expressions.dll` is around 5MB in size but since it also references to `System.Linq` (~400K) and `System.ObjectModel` (~80K) (which that are not referenced by STJ) it becomes ~5.5MB (crossgen'd size).
+  - STJ is considering removing its reference to using reflection emit, so referencing `System.Linq.Expressions.dll` would also require keeping the references to the various System.Reflection.* assemblies. However, since these are currently implemented in the runtime, it is not a factor today for disk size.
+
+- STJ is somewhat being pushed downstack with usage by other framework assemblies including the recent "BinaryData" work, so basic questions around the assembly dependency tree and potential cycles come into play.
+
+Thus the proposal is to add a `System.Text.Json.Dynamic.dll` assembly that will have a reference to both `System.Text.Json.dll` and `System.Linq.Expressions.dll` and contain the implementation supporting `dynamic`.
+
+An alternative which may or may not be feasible in the 6.0 timeframe is to extend the linker to remove dependencies on applications such as Blazor or a stand-alone app if the `dynamic` feature is not used. Currently the linker is only ran against framework assemblies, not application assemblies. This will help the stand-alone scenario and perhaps Blazor, but not the downstack concerns.
+
+_For naming, also consider `S.T.J.DynamicSerialization.dll`._
+
+A more open-ended name of `S.T.J.Extensions.dll` was considered but is not discoverable, so a more feature-specific name was considered a better choice. By going with feature-specific names like `S.T.J.Dynamic.dll`, we may end up with additional assemblies in the future. If the need arises, these extension assemblies could forward to private common implementation assembly(s).
+
+## Proposed API to enable dynamic mode
+In the new STJD assembly:
+```cs
+namespace System.Text.Json
+{
+    public static class JsonExtensions
+    {
+        public static void EnableDynamic(JsonSerializerOptions this options);
+    }
+}
+```
+
+Thus a consumer would reference STJD.dll and then author code such as:
+```cs
+    var options = new JsonSerializerOptions();
+    options.EnableDynamicTypes();
```

Comment:
What is going to be the default for ASP.NET ? Is ASP.NET going to enable the dynamic by default (and thus be fat out of the box), or not and be inconvenient to use out of the box? It is a very hard choice to make.

Reply:
IMO it should be opt-in since:
- It is faster, uses less private bytes and should have faster startup.
  - A rough benchmark shows that dynamic is ~2.6x slower when deserializing + serializing a smallish POCO with a few properties including a collection and nested object.
- Non-breaking: backwards compat for deserializing `System.Object` types as `JsonElement`.
- It is easy to opt-in.
- nit: if dynamic is enabled by default it is not easy to opt-out: you'd have to remove the custom converter from `options.Converters` although we could add a helper for that.

However, if it is common or useful enough (including compat with Newtonsoft) then opt-in could be considered. @pranavkm thoughts on how common the C# `dynamic` usage is?

Reply:
From the code I've encountered, it's far more common to see the use of `JObject` in application code over the use of `dynamic` both to read and build a DOM. I would have expected that having a editable DOM that's available by default would be something we would want to enable.

A few points against enabling `dynamic` by default:

* Globally changing JSON defaults in ASP.NET Core is fairly easy. We have documented how to enable or opt-out of defaults and users haven't had too many complaints.
* Generally any kind of opaque type (dynamic / JObject / JsonNode) makes it harder to document an API. For 6.0, one of our themes is to make APIs more descriptive, so encouraging this pattern is counter to goals. That said, I doubt there's a big overlap of users that use `dynamic` in their APIs and also have Swagger enabled.
* Undoing this option is apparently more tedious than enabling it. It would follow we make the easier choice and not enable it by default.

However the volume of support for the [issue](https://github.com/dotnet/runtime/issues/29690) makes me want to second guess myself. I'm just surprised how popular it is and I don't think I have enough data to unilaterally make a decision here. I'll follow up with @bradygaster to get you a decision. It's great we're having this discussion early, we have ample time to get the right defaults before 6.0 ships.

Reply:
I'm not a big fan of the dynamic support in Newtonsoft.Json. People want it to work in such a way that they just use .NET primitive types and everything just works. The reality is the fact you're using JValue/JObject/JArray leaks out.

Reply:
> I'm not a big fan of the dynamic support in Newtonsoft.Json. People want it to work in such a way that they just use .NET primitive types and everything just works.

Can you elaborate a bit? Both Newtonsoft and the design here support .NET primitive types when in "edit mode" meaning after deserialization or for a new object, the user can assign a raw integer, List, string, standard POCO type, etc. to a property on a dynamic object (or add an element to dynamic array) and that gets serialized that as expected. 

> The reality is the fact you're using JValue/JObject/JArray leaks out.

During deserialization, we need a type that implements `IDynamicMetaObjectProvider` thus we need to create JsonValue,JsonObject,JsonArray (or equivalent) on deserialization. Are you suggesting that the type implementing `IDynamicMetaObjectProvider` has no additional functionality or an internal implementation instead of JsonValue etc?